osu.Server.Spectator/Hubs/MultiplayerQueue.cs

```diff
@@ -0,0 +1,315 @@
+        /// Changes the queueing mode.
+        /// </summary>
+        /// <param name="newMode">The new mode.</param>
+        public async Task ChangeMode(QueueModes newMode)
+        {
+            if (mode == newMode)
+                return;
+
+            if (newMode == QueueModes.HostOnly)
+            {
+                // When changing to host-only mode, ensure that exactly one non-expired playlist item exists.
+                using (var db = dbFactory.GetInstance())
+                {
+                    // Remove all but the current and expired items. The current item may be re-used for host-only mode if it's non-expired.
+                    foreach (var item in await db.GetAllPlaylistItems(room.RoomID))
+                    {
+                        if (item.expired || item.id == room.Settings.PlaylistItemId)
+                            continue;
+
+                        await db.RemovePlaylistItemAsync(room.RoomID, item.id);
+                        await hub.OnPlaylistItemRemoved(room, item.id);
+                    }
+
+                    // Always ensure that at least one non-expired item exists by duplicating the current item if required.
+                    if (CurrentItem.Expired)
+                        await duplicateCurrentItem(db);
+                }
+            }
+
+            mode = newMode;
+
+            // When changing modes, items could have been added (above) or the queueing order could have change.
+            await updateCurrentItem();
+        }
+
+        /// <summary>
+        /// Expires the current playlist item and advances to the next one in the order defined by the queueing mode.
+        /// </summary>
+        public async Task FinishCurrentItem()
+        {
+            using (var db = dbFactory.GetInstance())
+            {
+                // Expire and let clients know that the current item has finished.
+                await db.ExpirePlaylistItemAsync(CurrentItem.ID);
+                CurrentItem.Expired = true;
+                await hub.OnPlaylistItemChanged(room, CurrentItem);
+
+                // In host-only mode, duplicate the playlist item for the next round.
+                if (mode == QueueModes.HostOnly)
+                    await duplicateCurrentItem(db);
+            }
+
+            await updateCurrentItem();
+        }
+
+        /// <summary>
+        /// Add a playlist item to the room's queue.
+        /// </summary>
+        /// <param name="item">The item to add.</param>
+        /// <param name="user">The user adding the item.</param>
+        /// <exception cref="NotHostException">If the adding user is not the host in host-only mode.</exception>
+        /// <exception cref="InvalidStateException">If the given playlist item is not valid.</exception>
+        public async Task AddItem(MultiplayerPlaylistItem item, MultiplayerRoomUser user)
+        {
+            if (mode == QueueModes.HostOnly && (room.Host == null || !user.Equals(room.Host)))
```

Comment:
What's the case of `null` host?

Reply:
This was to work around the issue of `MultiplayerRoomUser` implementing `Equals(X)` instead of `Equals(X?)`. I've fixed this osu!-side now, so removed the extra `null` check.

Do note, though, that `null` is the default value for host. Could also be null in the future with something like automated rooms.